### DISCUSSION TITLE
Selector failing to select a child element in a TOML file

### DISCUSSION THREAD

**Discussion by MrDionysus** [TomWright/dasel#97] [Q&A] [answered]
Thanks for this amazing tool!  I'm trying to search and replace in a TOML file, but I'm unable to select child elements from within the file - [config.toml.txt](https://github.com/TomWright/dasel/files/6038982/config.toml.txt)

Selecting the property "check_interval" works fine when I run
`dasel select -f config.toml -s "check_interval"`

But if I try to select disable_cache from [runners.docker], I run 
'dasel select -f config.toml -s "runners.docker.disable_cache"' 

and receive an error of 
'Error: could not query node: could not find value: selector [PROPERTY] does not support value: reflect.Kind: interface'

I've also tried "runners:docker:disable_cache" and  "runners.runners.docker.disable_cache", but those aren't working either.  
I assume I'm just messing up the syntax, but would appreciate any clarification.  This seems to be the only good parser in town for TOML, so some examples of TOML usage in the "Examples" section might help a lot.  

Thanks again, your work is greatly appreciated.

**Comment by TomWright** [comment, marked as answer]
Hey @MrDionysus,

I'm glad you're finding dasel useful.

I've taken a look and the issue is in your TOML file.

I see you have `[[runners]]` - notice the double square brackets. This causes runners to be an array containing a map with your expected values.

You can either use this selector: `.runners.[0].docker.disable_cache` - [example on playground](https://dasel.tomwright.me/s/1dee7674-d716-4ed6-a43f-407a99da0d4d)

Or change `[[runners]]` to `[runners]` and use this selector: `.runners.docker.disable_cache` - [example on playground](https://dasel.tomwright.me/s/5eeca682-12a5-4fec-8ac7-0cb879a77609)

I've also made a note to improve these error messages in the future.

**Reply by TomWright** [reply]
I've just pushed a change (although it's not released yet). The error output for what you were providing will look like this in future:

```
Error: could not query node: could not find value: selector [type:PROPERTY selector:.docker] does not support value: [kind:slice type:[]map[string]interface {}] ...
```

followed by the actual value of the node.

**Comment by MrDionysus** [comment]
Fantastic!  That helped a lot.

The only other problem I'm facing is using a put to add a list of strings instead of a single string.  Is that even possible?  For example, see below where there's a list of string values for MachineOptions:

  [runners.machine]
    IdleCount = 1
    IdleTime = 1800
    MaxBuilds = 10
    MachineDriver = "amazonec2"
    MachineName = "gitlab-docker-machine-%s"
    MachineOptions = [
      "amazonec2-access-key=XXXX",
      "amazonec2-secret-key=XXXX",
      "amazonec2-region=us-central-1",
      "amazonec2-vpc-id=vpc-xxxxx",
      "amazonec2-subnet-id=subnet-xxxxx",
      "amazonec2-zone=x",
      "amazonec2-use-private-address=true",
      "amazonec2-tags=runner-manager-name,gitlab-aws-autoscaler,gitlab,true,gitlab-runner-autoscale,true",
      "amazonec2-security-group=xxxxx",
      "amazonec2-instance-type=m4.2xlarge",
    ]

**Reply by TomWright** [reply]
You can use a `put document` command instead:

```
dasel put document -p toml -d json 'runners.machine.MachineOptions' '[
"amazonec2-access-key=XXXX",
"amazonec2-secret-key=XXXX",
"amazonec2-region=us-central-1",
"amazonec2-vpc-id=vpc-xxxxx",
"amazonec2-subnet-id=subnet-xxxxx",
"amazonec2-zone=x",
"amazonec2-use-private-address=true",
"amazonec2-tags=runner-manager-name,gitlab-aws-autoscaler,gitlab,true,gitlab-runner-autoscale,true",
"amazonec2-security-group=xxxxx",
"amazonec2-instance-type=m4.2xlarge"
]'
```

Here we are taking an JSON document containing an array of strings, inserting that into the correct place and then outputting the result as TOML again.

I had to define the document type as JSON because I don't know how to define an array without a key in TOML.
